From 8bdfefb4fd3de311294d9ae12df4fd84e64196ce Mon Sep 17 00:00:00 2001
From: Omri Itach <omrii@marvell.com>
Date: Mon, 25 Apr 2016 10:06:21 +0300
Subject: [PATCH 1102/1240] pinctrl: add support to keep default pin value

Added support to skip certain MPP pin default power-on settings.
Skip modifying default configuration by marking pin as 0xff in device tree.

This is required for A80x0 CP0 and CP1, which are connected using shared MPP pins:
CP0 use it's MPP[32-62] for shared pins with CP1 MPP[0-31].
shared MPPs must be left with their default power-on settings.

Change-Id: I8a63c12052ba7ad44ab18077dffec1de2ce7218c
Signed-off-by: Omri Itach <omrii@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/29286
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Haim Boot <hayim@marvell.com>
---
 doc/device-tree-bindings/misc/mvebu-pinctl.txt |  1 +
 drivers/gpio/mvebu_pinctl.c                    | 14 ++++++++++----
 2 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/doc/device-tree-bindings/misc/mvebu-pinctl.txt b/doc/device-tree-bindings/misc/mvebu-pinctl.txt
index 18b3720..c9d6981 100644
--- a/doc/device-tree-bindings/misc/mvebu-pinctl.txt
+++ b/doc/device-tree-bindings/misc/mvebu-pinctl.txt
@@ -24,6 +24,7 @@ SoC specific:
 Board specific:
 	- pin-func
 		Array specifying the function per pin, from left to right.
+		mark function as 0xff to keep default pin setting (skip modifying pin value)
 
 Example of a pinctl bank node:
 	pinctl@18000 {
diff --git a/drivers/gpio/mvebu_pinctl.c b/drivers/gpio/mvebu_pinctl.c
index 2644e3c..fb840e0 100644
--- a/drivers/gpio/mvebu_pinctl.c
+++ b/drivers/gpio/mvebu_pinctl.c
@@ -16,6 +16,7 @@
  * ***************************************************************************
  */
 
+/* #define DEBUG */
 #include <config.h>
 #include <common.h>
 #include <asm/system.h>
@@ -24,6 +25,7 @@
 #include <errno.h>
 #include <asm/arch-mvebu/fdt.h>
 #include <asm/arch/soc-info.h>
+#include <asm/arch-mvebu/mvebu.h>
 
 #define CONFIG_MAX_PINCTL_BANKS		4
 #define CONFIG_MAX_PINS_PER_BANK	100
@@ -132,6 +134,7 @@ int mvebu_pinctl_probe(void)
 	int count, i, err, pin;
 	u32 *pin_func;
 
+	debug_enter();
 
 	count = fdtdec_find_aliases_for_id(gd->fdt_blob, "pinctl",
 			COMPAT_MVEBU_PINCTL, node_list, CONFIG_MAX_PINCTL_BANKS);
@@ -173,11 +176,14 @@ int mvebu_pinctl_probe(void)
 		}
 
 		for (pin = 0; pin < pinctl->pin_cnt; pin++) {
-			err = pinctl_set_pin_func(i, pin, pin_func[pin]);
-			if (err)
-				printf("Warning: pin %d is not set for bank %d\n", pin, i);
+			if (pin_func[pin] != 0xff) {
+				err = pinctl_set_pin_func(i, pin, pin_func[pin]);
+				if (err)
+					printf("Warning: pin %d is not set for bank %d\n", pin, i);
+			} else
+				debug("Warning: pin %d value is not modified (kept as default\n", pin);
 		}
 	}
-
+	debug_exit();
 	return 0;
 }
-- 
1.9.1

